crates/storage/libmdbx-rs/src/txn_manager.rs

```diff
@@ -0,0 +1,331 @@
+                            self.add_aborted(ptr);
+
+                            // Abort the transaction
+                            let result = mdbx_result(unsafe { ffi::mdbx_txn_abort(ptr) });
+
+                            // Add the transaction to `aborted_active`. We can't remove it instantly
+                            // from the list of active transactions, because we iterate through it.
+                            aborted_active.push((ptr, duration, result.err()));
+                        }
+                    }
+
+                    // Walk through aborted transactions, and delete them from the list of active
+                    // transactions.
+                    for (ptr, open_duration, err) in aborted_active.iter().copied() {
+                        // Try deleting the transaction from the list of active transactions.
+                        let was_in_active = self.remove_active(ptr).is_some();
+                        if let Some(err) = err {
+                            // If there was an error when aborting the transaction, we need to
+                            // remove it from the list of aborted transactions, because otherwise it
+                            // will stay there forever.
+                            self.remove_aborted(ptr);
+                            if was_in_active && err != Error::BadSignature {
+                                // If the transaction was in the list of active transactions and the
+                                // error code is not `EBADSIGN`, then user didn't abort it.
+                                error!(target: "libmdbx", ?err, ?open_duration, "Failed to abort the long-lived read transactions");
+                            }
+                        } else {
+                            // Happy path, the transaction has been aborted by us with no errors.
+                            warn!(target: "libmdbx", ?open_duration, "Long-lived read transactions has been aborted");
+                        }
+                    }
+
+                    // Clear the list of aborted transactions, but not de-allocate the reserved
+                    // capacity to save on further pushes.
+                    aborted_active.clear();
+
+                    if !self.active.is_empty() || !self.aborted.is_empty() {
+                        trace!(
+                            target: "libmdbx",
+                            elapsed = ?now.elapsed(),
+                            active = ?self.active.iter().map(|entry| {
+                                let (ptr, start) = entry.pair();
+                                (*ptr, start.elapsed())
+                            }).collect::<Vec<_>>(),
+                            aborted = ?self.aborted.iter().map(|entry| *entry).collect::<Vec<_>>(),
+                            "Read transactions"
+                        );
+                    }
+
+                    std::thread::sleep(READ_TRANSACTIONS_CHECK_INTERVAL);
```

Comment:
hmm, is it reasonable to check this every 500ms if the default timeout is 5minutes?

perhaps we increase this a lot and for the edge case `max timeout + interval` we could find the lowest distance of all active transactions to their deadline and sleep for min(interval, deadline)

for example interval is 2 mins
but there's one transaction that's been open for 4min, we sleep for `min(5min - 4min, 2min)`

Reply:
this works only if new transactions aren't added to `active`, right?

In your example with 2 minutes, in the worst case, the transaction will be open for 2 minutes even though the max timeout is lower, because if there are no transactions, your expr will evaluate to `min(5min - 0, 2min) = 2min`. I like the idea of calculation using min, so we're waiting right until the next transaction needs to be killed, but I think the interval needs to stay at 500ms-10s.